### PR TITLE
Optimized ClusterStatsIndices to precomute shard stats

### LINKED ISSUE #14714: [BUG] Cluster Stats Request performs shard level aggregations on coordinator node instead of individual data nodes
### Describe the bug

The current approach for generating cluster stats indices incurs significant overhead in large clusters with numerous shards distributed across nodes. This issue arises due to the following steps:

1. Fetching Shard-Level Stats: Each node fetches a map containing individual shard statistics.
2. Accumulating on Request Node: The node receiving the REST request (request node) gathers this information from all participating nodes.
3. Inefficient Data Structures: Creating hash-maps from the StreamInput and iterating through them for accumulation are computationally expensive.

image

### Related component

Other

### To Reproduce

In TransportClusterStatsAction, each node is made to return ShardStats map as defined [here](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/action/admin/cluster/stats/TransportClusterStatsAction.java#L215) which is then iterated over in [ClusterStatsIndices](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/action/admin/cluster/stats/ClusterStatsIndices.java#L80-L107). This iteration goes through each shard stats entry in each node response.

### Expected behavior

The ShardLevel stats should be pre-computed in the data nodes prior to sending it back to the coordinator node. 

### Additional Details

**Plugins**
Please list all plugins currently enabled.

**Screenshots**

**Host/Environment (please complete the following information):**

**Additional context**

### PR DESCRIPTION
**Pull request by Pranshu-S**
### Description
The current approach for generating cluster stats indices incurs significant overhead in large clusters with numerous shards distributed across nodes. This issue arises due to the following steps:

1. Fetching Shard-Level Stats: Each node fetches a map containing individual shard statistics.
2. Accumulating on Request Node: The node receiving the REST request (request node) gathers this information from all participating nodes.
3. Inefficient Data Structures: Creating hash-maps from the StreamInput and iterating through them for accumulation are computationally expensive.

This pull request (PR) addresses this efficiency bottleneck by setting up each node to pre-compute its own statistics prior to sending the response. This allows the requested node to only perform node-level accumulation instead of the shard level. This is done by returning a new NodeIndexShard object containing the accumulated node-level statistics used by the request node for final aggregation.

The optimisation hides behind a flag - `useOptimizedClusterStatsResponse` (Defaults to False). This is only enabled in RestClusterStatsAction.

Improvement statistics:
Tested on a 20k shards (empty) cluster with 500 attempts and parallelisation of 5.

| Time taken for complete call | Flag Disabled | Flag Enabled | 
| --- | --- | --- |
| Average (s)	 | 0.77155 |	0.20277 |
| P90 (s)	| 1.04287	| 0.32985|
| Max (s)	| 1.50661	| 0.62719|
| Min (s)	| 0.26654	| 0.08716|
| |  image | image |

### Related Issues
Resolves #14714 

### Check List
- [x] Functionality includes testing.
- [ ] API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md), if applicable.
- [ ] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose), if applicable.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by shwetathareja** [comment]
@Pranshu-S can you also share memory overhead on the node serving the request should have reduced after the optimization

**Comment by Pranshu-S** [comment]
Testing on a large cluster yielded the following results - 

Current behavior
image
Optimisation results
image


| |	Without optimisation	| With optimisation |
-- | -- | --
Allocations in Heap in 100 calls serially | 321 GB | 30.44 GB
CPU Samples in ClusterStatsResponse.init | 68536 | 15007
CPU Samples in TransportClusterStatsAction.newNodeResponse | 84359 | 15009
Total Heap Size acumulated over 100 calls | 1.39TB | 885.53 GB

**Comment by Pranshu-S** [comment]
> ❌ Gradle check result for [aac6776](https://github.com/opensearch-project/OpenSearch/commit/aac6776628125e06164d72773c507fcac7ba7783): [FAILURE](https://build.ci.opensearch.org/job/gradle-check/42618/)
> 
> Please examine the workflow log, locate, and copy-paste the failure(s) below, then iterate to green. Is the failure [a flaky test](https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md#flaky-tests) unrelated to your change?

[Test Result](https://build.ci.opensearch.org/job/gradle-check/42618/testReport/) (1 failure / +1)
[org.opensearch.action.admin.indices.create.RemoteCloneIndexIT.testCreateCloneIndexFailure](https://build.ci.opensearch.org/job/gradle-check/42618/testReport/junit/org.opensearch.action.admin.indices.create/RemoteCloneIndexIT/testCreateCloneIndexFailure/)

Flaky - https://github.com/opensearch-project/OpenSearch/issues/14292

**Comment by Pranshu-S** [comment]
> ❌ Gradle check result for [79a7c19](https://github.com/opensearch-project/OpenSearch/commit/79a7c192820c7aef1abbfd0cf1c5a0ba894ee6e9):
> 
> Please examine the workflow log, locate, and copy-paste the failure(s) below, then iterate to green. Is the failure [a flaky test](https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md#flaky-tests) unrelated to your change?

[Test Result](https://build.ci.opensearch.org/job/gradle-check/42663/testReport/) (1 failure / ±0)
[org.opensearch.remotestore.RemoteStoreStatsIT.testDownloadStatsCorrectnessSinglePrimaryMultipleReplicaShards](https://build.ci.opensearch.org/job/gradle-check/42663/testReport/junit/org.opensearch.remotestore/RemoteStoreStatsIT/testDownloadStatsCorrectnessSinglePrimaryMultipleReplicaShards/)

Flaky - https://github.com/opensearch-project/OpenSearch/issues/14310

File: server/src/main/java/org/opensearch/action/admin/cluster/stats/ClusterStatsIndices.java [hunk 1]
```diff
@@ -202,6 +223,12 @@ public static class ShardStats implements ToXContentFragment {
 
         public ShardStats() {}
 
+        public ShardStats(StreamInput in) throws IOException {
```
Comment (shwetathareja) [inline server/src/main/java/org/opensearch/action/admin/cluster/stats/ClusterStatsIndices.java:226]: This was an internal object in ClusterStatsIndices, why are adding serialization to it.

File: server/src/main/java/org/opensearch/action/admin/cluster/stats/ClusterStatsNodeResponse.java [hunk 2]
```diff
@@ -116,6 +145,16 @@ public void writeTo(StreamOutput out) throws IOException {
...
         nodeInfo.writeTo(out);
         nodeStats.writeTo(out);
-        out.writeArray(shardsStats);
+        if (out.getVersion().onOrAfter(Version.V_3_0_0)) {
+            if (aggregatedNodeLevelStats != null) {
+                out.writeOptionalArray(null);
+                out.writeOptionalWriteable(aggregatedNodeLevelStats);
+            } else {
+                out.writeOptionalArray(shardsStats);
+                out.writeOptionalWriteable(null);
```
Comment (shwetathareja) [inline server/src/main/java/org/opensearch/action/admin/cluster/stats/ClusterStatsNodeResponse.java:154]: why are you writing this?

File: server/src/main/java/org/opensearch/rest/action/admin/cluster/RestClusterStatsAction.java [hunk 3]
```diff
@@ -66,6 +66,7 @@ public String getName() {
     public RestChannelConsumer prepareRequest(final RestRequest request, final NodeClient client) throws IOException {
         ClusterStatsRequest clusterStatsRequest = new ClusterStatsRequest().nodesIds(request.paramAsStringArray("nodeId", null));
         clusterStatsRequest.timeout(request.param("timeout"));
+        clusterStatsRequest.useOptimizedClusterStatsResponse(true);
```
[the remark this answers is not in the document]
Comment (Pranshu-S) [reply server/src/main/java/org/opensearch/rest/action/admin/cluster/RestClusterStatsAction.java]: During version upgrade, this boolean will not be propagated to the target node - [Ref](https://github.com/opensearch-project/OpenSearch/pull/14426/files#diff-266bdfbfa4bb22a2c8e119b254f99db4bb62a7c3faf5ee37b5fa65bccfb28ce3R53-R55). So for nodes with level below the new version, they will receive a usual `_cluster/stats` request and hence not cause ser/de issues

File: server/src/main/java/org/opensearch/action/admin/cluster/stats/TransportClusterStatsAction.java [hunk 4]
```diff
@@ -212,8 +212,14 @@ protected ClusterStatsNodeResponse nodeOperation(ClusterStatsNodeRequest nodeReq
...
 
-        return new ClusterStatsNodeResponse(nodeInfo.getNode(), clusterStatus, nodeInfo, nodeStats, shardsStats.toArray(new ShardStats[0]));
-
+        return new ClusterStatsNodeResponse(
+            nodeInfo.getNode(),
+            clusterStatus,
+            nodeInfo,
+            nodeStats,
+            shardsStats.toArray(new ShardStats[0]),
+            nodeRequest.request.useOptimizedClusterStatsResponse()
```
[the remark this answers is not in the document]
Comment (Pranshu-S) [reply server/src/main/java/org/opensearch/action/admin/cluster/stats/TransportClusterStatsAction.java]: Using this variable, we decide to either compute aggregated node level stats or use the default response. We first initialize the values here - [Ref](https://github.com/opensearch-project/OpenSearch/pull/14426/files#diff-dcfa6989665ff47faee99c72d630f5bc4d8452f160fe9220f9aa75f8126959dbR91-R106) and then later perform the ser/de [here](https://github.com/opensearch-project/OpenSearch/pull/14426/files#diff-dcfa6989665ff47faee99c72d630f5bc4d8452f160fe9220f9aa75f8126959dbR148-R155)

In summary, we use the boolean from the request to determine what to send in the response.

File: server/src/main/java/org/opensearch/action/admin/cluster/stats/ClusterStatsRequest.java [hunk 5]
```diff
@@ -49,8 +50,13 @@ public class ClusterStatsRequest extends BaseNodesRequest<ClusterStatsRequest> {
 
     public ClusterStatsRequest(StreamInput in) throws IOException {
         super(in);
+        if (in.getVersion().onOrAfter(Version.V_3_0_0)) {
+            useOptimizedClusterStatsResponse = in.readOptionalBoolean();
+        }
     }
 
+    private Boolean useOptimizedClusterStatsResponse = false;
```
[the remark this answers is not in the document]
Comment (Pranshu-S) [reply server/src/main/java/org/opensearch/action/admin/cluster/stats/ClusterStatsRequest.java]: Named it earlier to `aggregatedNodeLevelResponses`. Do we still want a refactor to `preAggregateStats`?

File: server/src/main/java/org/opensearch/action/admin/cluster/stats/ClusterStatsNodeResponse.java [hunk 6]
```diff
@@ -64,7 +66,12 @@ public ClusterStatsNodeResponse(StreamInput in) throws IOException {
         }
         this.nodeInfo = new NodeInfo(in);
         this.nodeStats = new NodeStats(in);
-        shardsStats = in.readArray(ShardStats::new, ShardStats[]::new);
+        if (in.getVersion().onOrAfter(Version.V_3_0_0)) {
+            this.shardsStats = in.readOptionalArray(ShardStats::new, ShardStats[]::new);
```
[the remark this answers is not in the document]
Comment (Pranshu-S) [reply server/src/main/java/org/opensearch/action/admin/cluster/stats/ClusterStatsNodeResponse.java:70]: I have modified the IO to send `optionalArray` instead of a concrete `array` using `readOptionalArray` and `writeOptionalWriteable` instead of `readArray` and `writeArray`.

I did this because we normally always expect this to present. Since now they can be null too, I modified it accordingly. Do you require me to revert back?

File: server/src/main/java/org/opensearch/action/admin/cluster/stats/ClusterStatsIndices.java (hunk 1 above)
Comment (Pranshu-S) [reply server/src/main/java/org/opensearch/action/admin/cluster/stats/ClusterStatsIndices.java:226]: This is done since we are now precomputing and storing the node level aggregated stats in the form of these ShardStats. 

This table can explain the pre-computation - 

| Before | After |
| --- | --- | 
| On each node, `indices.stats.ShardStats` is fetched from the indices service which contains stats of each shard on the node | On each node, ShardStats is fetched from the indices service which contains stats of each shard on the node |
| We return this list of `indices.stats.ShardStats` to the coordinator node | We iterate over the `indices.stats.ShardStats` returned from IndicesService and create `ClusterIndicesStats.ShardStats` for that specific node. We now create a hashmap of index_name to `ClusterIndicesStats.ShardStats` at a node level. This is then returned to the coordinator node |
| Coordinator node iterates over each `indices.stats.ShardStats` from each node response and creates a map of index_name to `ClusterIndicesStats.ShardStats`, i.e., for each index, the `indices.stats.ShardStats` returned from the nodes are iterated and added into the respective `ClusterIndicesStats.ShardStats`. In the end, the `countsPerIndex` map contains indices and their accumulated `ClusterIndicesStats.ShardStats` made by adding up values from the `indices.stats.ShardStats` from each node | The coordinator node iterates over each Node response and merges the `countsPerIndex` map from each node to make a final stats result. |


Since we are now creating `ClusterIndicesStats.ShardStats` on each node, we require this to be capable of serialising and deserialising.

File: server/src/main/java/org/opensearch/action/admin/cluster/stats/ClusterStatsNodeResponse.java (hunk 2 above)
Comment (Pranshu-S) [reply server/src/main/java/org/opensearch/action/admin/cluster/stats/ClusterStatsNodeResponse.java:154]: The decision to return aggregatedNodeLevelStats determines whether to return the aggregated information or the whole list of ShardStats as a whole. We only create the aggregated stats during initialisation [here](https://github.com/opensearch-project/OpenSearch/blob/10a78f7476f41a53dba3a05bb473b61192872ea0/server/src/main/java/org/opensearch/action/admin/cluster/stats/ClusterStatsNodeResponse.java#L102-L104)

Writing shardsStats only when the request comes in with aggregatedNodeLevelStats set to `false` implying `aggregatedNodeLevelStats` is false.

File: server/src/main/java/org/opensearch/action/admin/cluster/stats/AggregatedNodeLevelStats.java [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+import org.opensearch.search.suggest.completion.CompletionStats;
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Node level statistics used for ClusterStatsIndices for _cluster/stats call.
+ */
+public class AggregatedNodeLevelStats extends BaseNodeResponse {
```
Comment (Pranshu-S) [inline server/src/main/java/org/opensearch/action/admin/cluster/stats/AggregatedNodeLevelStats.java:31]: Default behaviour is as follows:

Coordinator node iterates over each `indices.stats.ShardStats` from each node response and creates a map of index_name to `ClusterIndicesStats.ShardStats`, i.e., for each index, the `indices.stats.ShardStats` returned from the nodes are iterated upon and added into the respective `ClusterIndicesStats.ShardStats` for that index. In the end, the `countsPerIndex` map contains indices and their accumulated `ClusterIndicesStats.ShardStats` made by adding up values from the indices.stats.ShardStats from each node

As this adds an additional overhead on the coordinating node of going through each `indices.stats.ShardStats` returned from each Node, we optimise it by computing `AggregatedNodeLevelStats` on each node and returning this itself.

File: server/src/main/java/org/opensearch/action/admin/cluster/stats/ClusterStatsIndices.java [hunk 8]
```diff
@@ -78,26 +81,44 @@ public ClusterStatsIndices(List<ClusterStatsNodeResponse> nodeResponses, Mapping
...
-                indexShardStats.total++;
-
-                CommonStats shardCommonStats = shardStats.getStats();
-
-                if (shardStats.getShardRouting().primary()) {
-                    indexShardStats.primaries++;
-                    docs.add(shardCommonStats.docs);
+            // Aggregated response from the node
+            if (r.getAggregatedNodeLevelStats() != null) {
+                r.getAggregatedNodeLevelStats().indexStatsMap.forEach(
```
Comment (Pranshu-S) [inline server/src/main/java/org/opensearch/action/admin/cluster/stats/ClusterStatsIndices.java:86]: In ClusterStatsIndices, we create a `countsPerIndex` which has a cluster level accumulated information on each index and its ShardStats (accumulated over each Shard from each node belonging to that index)

To handle multi-version scenarios where some nodes in older versions may not have the `AggregatedNodeLevelStats` in their response we do the following:

1. If we do have `AggregatedNodeLevelStats`, we merge the results into the overall `countsPerIndex` map. This is done so that stats of the same index from different nodes are added up.
2. If we do not have `AggregatedNodeLevelStats`, we use the default method of adding the shardStats by iterating over each `indices.stats.ShardStats` returned from each node.

File: server/src/main/java/org/opensearch/action/admin/cluster/stats/ClusterStatsIndices.java [hunk 9]
```diff
@@ -185,11 +206,11 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
      * @opensearch.api
      */
     @PublicApi(since = "1.0.0")
-    public static class ShardStats implements ToXContentFragment {
+    public static class ShardStats implements ToXContentFragment, Writeable {
```
Comment (Pranshu-S) [inline server/src/main/java/org/opensearch/action/admin/cluster/stats/ClusterStatsIndices.java:209]: Making this Writeable as we create this in `AggregatedNodeLevelStats` on each node and return it from there. The coordinating node will then just merge the results from each node instead of going over each `indices.stats.ShardStats`.

As this now comes in the ser/de path, needs to made Writeable.

File: server/src/main/java/org/opensearch/action/admin/cluster/stats/ClusterStatsNodeResponse.java [hunk 10]
```diff
@@ -81,6 +88,24 @@ public ClusterStatsNodeResponse(
...
+        @Nullable ClusterHealthStatus clusterStatus,
+        NodeInfo nodeInfo,
+        NodeStats nodeStats,
+        ShardStats[] shardsStats,
+        boolean useAggregatedNodeLevelResponses
+    ) {
+        super(node);
+        this.nodeInfo = nodeInfo;
+        this.nodeStats = nodeStats;
+        if (useAggregatedNodeLevelResponses) {
```
Comment (Pranshu-S) [inline server/src/main/java/org/opensearch/action/admin/cluster/stats/ClusterStatsNodeResponse.java:102]: We only create `aggregatedNodeLevelStats` when the param is passed. Else we do not. This reduces the request size when not required.

File: server/src/main/java/org/opensearch/action/admin/cluster/stats/ClusterStatsRequest.java [hunk 11]
```diff
@@ -49,8 +50,13 @@ public class ClusterStatsRequest extends BaseNodesRequest<ClusterStatsRequest> {
 
     public ClusterStatsRequest(StreamInput in) throws IOException {
         super(in);
+        if (in.getVersion().onOrAfter(Version.V_3_0_0)) {
```
Comment (Pranshu-S) [inline server/src/main/java/org/opensearch/action/admin/cluster/stats/ClusterStatsRequest.java:53]: We only send and recieve this on new versions, this is not sent in old versions as this may end up breaking during version upgrades.

**Comment by Pranshu-S** [comment]
> ❌ Gradle check result for [31fd6d8](https://github.com/opensearch-project/OpenSearch/commit/31fd6d86266ca15bae3d70ae538623897eb88f44): [FAILURE](https://build.ci.opensearch.org/job/gradle-check/42916/)
> 
> Please examine the workflow log, locate, and copy-paste the failure(s) below, then iterate to green. Is the failure [a flaky test](https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md#flaky-tests) unrelated to your change?

[Test Result](https://build.ci.opensearch.org/job/gradle-check/42916/testReport/) (2 failures / +1)
[org.opensearch.action.admin.indices.create.RemoteSplitIndexIT.classMethod](https://build.ci.opensearch.org/job/gradle-check/42916/testReport/junit/org.opensearch.action.admin.indices.create/RemoteSplitIndexIT/classMethod/)
[org.opensearch.action.admin.cluster.node.tasks.ResourceAwareTasksTests.testTaskResourceTrackingDuringTaskCancellation](https://build.ci.opensearch.org/job/gradle-check/42916/testReport/junit/org.opensearch.action.admin.cluster.node.tasks/ResourceAwareTasksTests/testTaskResourceTrackingDuringTaskCancellation/)


Flaky - https://github.com/opensearch-project/OpenSearch/issues/14296, https://github.com/opensearch-project/OpenSearch/issues/14293

**Comment by dblock** [comment]
Maybe someone more familiar with this are (@shwetathareja or @Bukhtawar?) can please CR this?

File: server/src/main/java/org/opensearch/action/admin/cluster/stats/ClusterStatsNodeResponse.java (hunk 10 above)
Comment (shwetathareja) [reply server/src/main/java/org/opensearch/action/admin/cluster/stats/ClusterStatsNodeResponse.java:102]: I am still unclear why you need useAggregatedNodeLevelResponses boolean in response as well. If recieved it in request, you will send the response accordingly.

File: server/src/main/java/org/opensearch/action/admin/cluster/stats/AggregatedNodeLevelStats.java [hunk 12]
```diff
@@ -0,0 +1,98 @@
...
+
+    DocsStats docs;
+    StoreStats store;
+    FieldDataStats fieldData;
+    QueryCacheStats queryCache;
+    CompletionStats completion;
+    SegmentsStats segments;
```
[the remark this answers is not in the document]
Comment (Pranshu-S) [reply server/src/main/java/org/opensearch/action/admin/cluster/stats/AggregatedNodeLevelStats.java]: Refactored to use CommonStats instead.

File: server/src/main/java/org/opensearch/action/admin/cluster/stats/AggregatedNodeLevelStats.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (Pranshu-S) [reply server/src/main/java/org/opensearch/action/admin/cluster/stats/AggregatedNodeLevelStats.java]: I have made it a subclass of `ClusterStatsNodeResponse`

File: server/src/main/java/org/opensearch/action/admin/cluster/stats/ClusterStatsNodeResponse.java (hunk 10 above)
Comment (Pranshu-S) [reply server/src/main/java/org/opensearch/action/admin/cluster/stats/ClusterStatsNodeResponse.java:111]: We do not use the flag in the response. It is just used for generating the response and is not passed around.

File: server/src/main/java/org/opensearch/action/admin/indices/stats/CommonStats.java [hunk 13]
```diff
@@ -544,4 +544,29 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
         }
         return builder;
     }
+
+    /**
+     * Node level statistics used for ClusterStatsIndices for _cluster/stats call.
+     */
+    @PublicApi(since = "2.0.0")
```
Comment (shwetathareja) [inline server/src/main/java/org/opensearch/action/admin/indices/stats/CommonStats.java:551]: this should be 2.16.0 right

**Comment by Pranshu-S** [comment]
[Test Result](https://build.ci.opensearch.org/job/gradle-check/43164/testReport/) (2 failures / +2)
[org.opensearch.cluster.MinimumClusterManagerNodesIT.testThreeNodesNoClusterManagerBlock](https://build.ci.opensearch.org/job/gradle-check/43164/testReport/junit/org.opensearch.cluster/MinimumClusterManagerNodesIT/testThreeNodesNoClusterManagerBlock/)
[org.opensearch.remotestore.CreateRemoteIndexClusterDefaultDocRepIT.testDefaultRemoteStoreNoUserOverrideExceptReplicationTypeSegment](https://build.ci.opensearch.org/job/gradle-check/43164/testReport/junit/org.opensearch.remotestore/CreateRemoteIndexClusterDefaultDocRepIT/testDefaultRemoteStoreNoUserOverrideExceptReplicationTypeSegment/)

Flaky behaviour - 
CreateRemoteIndexClusterDefaultDocRepIT.testDefaultRemoteStoreNoUserOverrideExceptReplicationTypeSegment
```
com.carrotsearch.randomizedtesting.UncaughtExceptionError: Captured an uncaught exception in thread: Thread[id=4922, name=Thread-71, state=RUNNABLE, group=TGRP-CreateRemoteIndexClusterDefaultDocRepIT]
```

Similar to tests in - https://github.com/opensearch-project/OpenSearch/issues/14381

---
org.opensearch.cluster.MinimumClusterManagerNodesIT.testThreeNodesNoClusterManagerBlock
Flaky - https://github.com/opensearch-project/OpenSearch/issues/14289

**Review by shwetathareja — APPROVED** [review]
